### PR TITLE
[GR-36425] Always execute assumption expressions instead of caching if they can be proven PE constant.

### PR DESCRIPTION
**Pull request by graalvmbot**
* Fix specialization statistics test to not fail if run on Eclipse.
* Always execute assumption expressions instead of caching if they can be proven PE constant.

### CODE REVIEW

File: truffle/src/com.oracle.truffle.api.dsl.test/src/com/oracle/truffle/api/dsl/test/GenerateUncachedTest.java [hunk 1]
```diff
@@ -209,7 +209,7 @@ public void testUncached5() {
         Uncached5Node.testAssumption.invalidate();
         assertEquals("s2", node.execute(42));
         Uncached5Node.testAssumption = null;
-        assertEquals("s1", node.execute(42));
+        assertEquals("s2", node.execute(42));
```
Comment (chumer) [inline truffle/src/com.oracle.truffle.api.dsl.test/src/com/oracle/truffle/api/dsl/test/GenerateUncachedTest.java:212]: This was a bug I fixed on the way. Uncached behavior was inconsistent with cached behavior.
In cached behavior it was not activating the specialization if the assumption was null, but in uncached behavior it did.
This was an bug caused by the fallback behavior where you need to check for null.
I don't expect this to have a big impact, as null assumption expressions are likely rare.

**Review by smarr — COMMENTED** [review]
Looks good to me.
Seems to do what I'd expect it to, when tested on TruffleRuby.

Though, I am not 100% sure I read the tests right, would perhaps the following still be useful?
- interaction with data classes (i.e. when `@Cached` is used, where assumptions are stored in fields)
- that methods are indeed not considered
- that a non-final/non-`@CompilationFinal` field not being used

**Comment by chumer** [comment]
> * interaction with data classes (i.e. when `@Cached` is used, where assumptions are stored in fields)

There are tests for this here: https://github.com/oracle/graal/pull/4247/files#diff-4338bbbb748949f0e843a04a957b4d4c15ec3e1747497e918c27e78076b048bcR690
No?

> * that methods are indeed not considered
> * that a non-final/non-`@CompilationFinal` field not being used

There were already tests for such cases in AssumptionsTest afaics.
But maybe a more explicit test for could not hurt. Will add.

File: truffle/CHANGELOG.md [hunk 2]
```diff
@@ -11,8 +11,7 @@ This changelog summarizes major changes between Truffle versions relevant to lan
...
 * GR-10128 Added the `usageSyntax` element to the [Option](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/Option.html#usageSyntax--) allowing developers to specify the syntax their option expects. See the javadoc for more information.
-* GR-10128 Added the `website` property to the [TruffleLanguage.Registration](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/TruffleLanguage.Registration.html#website--) and [TruffleInstrument.Registration](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/instrumentation/TruffleInstrument.Registration.html#website--) allowing language and instrument developers to specify a URL for a web site with further information about their language/tool.
```
[the remark this answers is not in the document]
Comment (chumer) [reply truffle/CHANGELOG.md:14]: That is just a changelog duplication fix. That sometimes happens on merge conflicts with merge mode union on files (see https://github.com/oracle/graal/blob/master/.gitattributes#L1). That is better than dealing with merge conflicts in the file all the time.

**Comment by smarr** [comment]
> There are tests for this here

Ah, I missed those.

**Pull request merged**